src/Sylius/Bundle/LocaleBundle/Form/DataTransformer/ArrayGregorianToCalendarSystemTransformer.php

```diff
@@ -0,0 +1,92 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Bundle\LocaleBundle\Form\DataTransformer;
+
+use Sylius\Bundle\LocaleBundle\Templating\Helper\LocaleHelper;
+use Sylius\Component\Locale\Calendars;
+use Symfony\Component\Form\DataTransformerInterface;
+use Symfony\Component\Form\Exception\TransformationFailedException;
+
+/**
+ * @author Aram Alipoor <aram.alipoor@gmail.com>
+ */
+class ArrayGregorianToCalendarSystemTransformer implements DataTransformerInterface
+{
+    /**
+     * @var LocaleHelper
+     */
+    private $localeHelper;
+
+    /**
+     * @param $localeHelper
+     */
+    public function __construct(LocaleHelper $localeHelper)
+    {
+        $this->localeHelper = $localeHelper;
+    }
+
+    /**
+     * @param array $date Normalized date array.
+     *
+     * @return array Localized calendar-aware date array.
+     */
+    public function transform($date)
+    {
+        if (!is_array($date)) {
+            throw new TransformationFailedException('Expected an array.');
+        }
+
+        if (Calendars::GREGORIAN === $this->localeHelper->getCurrentCalendar()) {
+            return $date;
+        }
+
+        $dateTime = \DateTime::createFromFormat('Y-m-d', $date['year'] . '-' . $date['month'] . '-' . $date['day']);
+        $translatedDate = $this->localeHelper->formatDate($dateTime, 'Y-m-d');
+
+        list($year, $month, $day) = explode('-', $translatedDate);
+
+        return array(
+            'year' => $year,
+            'month' => $month,
+            'day' => $day
+        );
+    }
+
+    /**
+     * @param array $value Localized calendar-aware date array
+     *
+     * @return array Normalized date
+     *
+     * @throws TransformationFailedException If the given value is not an array.
+     */
+    public function reverseTransform($value)
```

Comment:
You could add ``array`` keyword before ``$value`` to ensure, that passed value is array. Thanks to that, you'll be able to remove first condition below.